### PR TITLE
caching failing test

### PR DESCRIPTION
**Pull request by mssimi**

### CODE REVIEW

**Comment by TomasVotruba** [comment]
Thanks! What needs to be done to fix it?

**Comment by mssimi** [comment]
Every single time I update rector something does not work, I have to spend so much time investigating what is wrong, you asked for failing test, now you got it, fix it or not, if not I will keep using half year old rector. I wanted to use rector to spend less time refactoring, but this way it actually does opposite. I know it is open source and I don't blame anyone am just saying.

**Comment by TomasVotruba** [comment]
No problem. This is rather edge case that would greatly benefit from research you've done so far :+1: 
We're working hard on other issues. It's team work effort that improves the code base.

Contribution is voluntary of course :)

**Comment by mssimi** [comment]
Of course, if not working cache is edge case. :+1:

**Comment by mssimi** [comment]
Maybe test was misunderstood but single one cache clean collision in any file analyzed causes rector config cache is invalidated, thus no caching is applied at all.

**Comment by staabm** [comment]
So you are saying, because of this cache collisions rector is not usable for you? What/When does not work because of this overly aggressive cache cleaning?

**Comment by mssimi** [comment]
I have large code base, without cache rector is processing 5 mins and I and my team mates do processing dozens of times everyday, so yes for us it is unusable. On other hand if you or anyone uses refactor for one time refactorings, than you are not affected much.

File: packages-tests/Caching/Cache/CacheTest.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        $this->cache->save('configuration_hash_srv-rector-php', CacheKey::CONFIGURATION_HASH_KEY, 'something');
+        $this->assertNotNull(
+            $this->cache->load('configuration_hash_srv-rector-php', CacheKey::CONFIGURATION_HASH_KEY)
+        );
+
+        $this->cache->save('fc9a36aa55dbb3d22e66e71d4f824cf3f8aba88d', CacheKey::CONFIGURATION_HASH_KEY, 'anything');
+        $this->cache->clean('fc9a36aa55dbb3d22e66e71d4f824cf3f8aba88d');
+
+        $this->assertNotNull( // clean actually also cleans other cache key
+            $this->cache->load('configuration_hash_srv-rector-php', CacheKey::CONFIGURATION_HASH_KEY)
```
Comment (staabm) [inline packages-tests/Caching/Cache/CacheTest.php]: Not sure this assumption makes sense: when working with a cache you always need to be prepared for the situation that the cache cannot deliver the key you ask for.

in this example its a bit suprising, that because of impl details the key cannot be delivered but I think it might be a issue from the calling code that this gets a real problem.

therefore I asked for the whole picture, when this scenario actually creates a problem for you

**Comment by mssimi** [comment]
I mean if we in team run rector 50 times per day, then it costs us 50*5 mins, which we cannot afford.

File: packages-tests/Caching/Cache/CacheTest.php (hunk 1 above)
Comment (mssimi) [reply packages-tests/Caching/Cache/CacheTest.php]: Ok, sorry. I can give real case I am experiencing. I have code base, where I scan about 5000 files and about 1% of them collide with `configuration_hash_srv-rector-php` which is cache key for rector's config file, if this key gets invalidated, no caching is applied so effectivelly caching is not working for us at all.

Comment (staabm) [reply packages-tests/Caching/Cache/CacheTest.php]: Shot in the dark: As a workaround you could try changing

https://github.com/rectorphp/rector-src/blob/6238705e938a643cb84e0b4275cb682d26307975/packages/Caching/ValueObject/Storage/FileCacheStorage.php#L86-L90

to

```php
$this->smartFileSystem->remove([
            $cacheFilePaths->getFilePath(),
        ]);
```

this should fix the problem I guess, but will leave empty directories in your cache folders after cleanup

**Comment by staabm** [comment]
> Ok, sorry. I can give real case I am experiencing. I have code base, where I scan about 5000 files and about 1% of them collide with configuration_hash_srv-rector-php which is cache key for rector's config file, if this key gets invalidated, no caching is applied so effectivelly caching is not working for us at all.

now we are talking. This makes a lot more sense now 👍

File: packages-tests/Caching/Cache/CacheTest.php (hunk 1 above)
Comment (mssimi) [reply packages-tests/Caching/Cache/CacheTest.php]: I can confirm this works.

Comment (staabm) [reply packages-tests/Caching/Cache/CacheTest.php]: I dont know the details/background of the caching, just made a guess regarding the above fix/workaround. Maybe @TomasVotruba can tell

Comment (mssimi) [reply packages-tests/Caching/Cache/CacheTest.php]: @TomasVotruba So can we keep these folders?

Comment (TomasVotruba) [reply packages-tests/Caching/Cache/CacheTest.php]: I have barely any knowledge about caching system, so I can't help here. It's based on implementation in PHPStan.

Comment (staabm) [reply packages-tests/Caching/Cache/CacheTest.php]: @mssimi i sent a PR to your fork which might fix the problem while preserving the old behavior

Comment (mssimi) [reply packages-tests/Caching/Cache/CacheTest.php]: @staabm Sorry, missed that. :)

Comment (mssimi) [reply packages-tests/Caching/Cache/CacheTest.php]: @staabm can we skip that phpstan error?

Comment (staabm) [reply packages-tests/Caching/Cache/CacheTest.php]: @mssimi I intentionally used rmdir() instead of the filesystem adapter because we only want to delete the directory if it is empty. Did you verify the fix works for you?

@TomasVotruba do you agree adding these errors to the ignore list is acceptable? The fix adresses the initially failling test of this PR

File: packages/Caching/ValueObject/Storage/FileCacheStorage.php [hunk 2]
```diff
@@ -84,10 +84,11 @@ public function clean(string $cacheKey): void
         $cacheFilePaths = $this->getCacheFilePaths($cacheKey);
 
         $this->smartFileSystem->remove([
-            $cacheFilePaths->getFirstDirectory(),
-            $cacheFilePaths->getSecondDirectory(),
             $cacheFilePaths->getFilePath(),
         ]);
+        // remove dirs only if empty to prevent over aggressive delete
+        @rmdir($cacheFilePaths->getSecondDirectory());
+        @rmdir($cacheFilePaths->getFirstDirectory());
```
Comment (TomasVotruba) [inline packages/Caching/ValueObject/Storage/FileCacheStorage.php]: Better avoid `@` magic, that will bite us sooner or later.

`removeIfEmpty()` would tell the same without comment

File: packages-tests/Caching/Cache/CacheTest.php (hunk 1 above)
Comment (mssimi) [reply packages-tests/Caching/Cache/CacheTest.php]: @staabm Yep I understand, but I don't know, if Tomas let us ignore that phpstan thing :) I can confirm it fixes my problem.

File: packages/Caching/ValueObject/Storage/FileCacheStorage.php (hunk 2 above)
Comment (mssimi) [reply packages/Caching/ValueObject/Storage/FileCacheStorage.php]: @staabm In that case we cannot use rmdir and glob or scandir must be used.

Comment (mssimi) [reply packages/Caching/ValueObject/Storage/FileCacheStorage.php]: I have checked phpstan caching and they don't remove that at all, they just returns invalid cache and later override it.

Comment (mssimi) [reply packages/Caching/ValueObject/Storage/FileCacheStorage.php]: I think it does not make sense to remove that folders, we got that file, which exist in code base and it is invalid and we are sure that cache file will be created again with new checksum, at the same place, there is totally no need for removing that folders.

Comment (staabm) [reply packages/Caching/ValueObject/Storage/FileCacheStorage.php]: Ok fine, kill it :-)

Comment (mssimi) [reply packages/Caching/ValueObject/Storage/FileCacheStorage.php]: Ok, I have removed single cache clean completely, since it is not needed, invalid cache files are overridden later in code, just like phpstan do that.

File: packages-tests/Caching/Detector/ChangedFilesDetectorTest.php [hunk 3]
```diff
@@ -33,9 +33,6 @@ public function testHasFileChanged(): void
         $this->changedFilesDetector->addFileWithDependencies($smartFileInfo, []);
 
         $this->assertFalse($this->changedFilesDetector->hasFileChanged($smartFileInfo));
-        $this->changedFilesDetector->invalidateFile($smartFileInfo);
-
-        $this->assertTrue($this->changedFilesDetector->hasFileChanged($smartFileInfo));
```
Comment (TomasVotruba) [inline packages-tests/Caching/Detector/ChangedFilesDetectorTest.php:38]: This seems to kill the original purpose of the test. 
It should be kept, even if with different code.

File: packages/Caching/Detector/ChangedFilesDetector.php [hunk 4]
```diff
@@ -45,12 +45,6 @@ public function hasFileChanged(SmartFileInfo $smartFileInfo): bool
         return $currentFileHash !== $cachedValue;
     }
 
-    public function invalidateFile(SmartFileInfo $smartFileInfo): void
-    {
-        $fileInfoCacheKey = $this->getFileInfoCacheKey($smartFileInfo);
-        $this->cache->clean($fileInfoCacheKey);
-    }
-
```
Comment (TomasVotruba) [inline packages/Caching/Detector/ChangedFilesDetector.php:53]: How is the file invalidated now?

File: packages/Caching/UnchangedFilesFilter.php [hunk 5]
```diff
@@ -29,7 +29,6 @@ public function filterAndJoinWithDependentFileInfos(array $fileInfos): array
             }
 
             $changedFileInfos[] = $fileInfo;
-            $this->changedFilesDetector->invalidateFile($fileInfo);
 
             $dependentFileInfos = array_merge(
```
Comment (TomasVotruba) [inline packages/Caching/UnchangedFilesFilter.php:33]: Where is this moved?

File: src/Console/Command/ProcessCommand.php [hunk 6]
```diff
@@ -135,9 +135,6 @@ protected function execute(InputInterface $input, OutputInterface $output): int
         $processResult = $this->processResultFactory->create($files);
         $outputFormatter->report($processResult, $configuration);
 
-        // invalidate affected files
-        $this->invalidateCacheChangedFiles($processResult);
-
```
Comment (TomasVotruba) [inline src/Console/Command/ProcessCommand.php:140]: Where are the changed files invalidated now?

**Review by TomasVotruba — APPROVED** [review]
Removing code without adding test looks very suspicious. It's like removing security layer and all it's tests :) 

Having 2-3 new tests would make this much easier to review.

**Review by TomasVotruba — CHANGES_REQUESTED** [review]
Re-click

File: src/Console/Command/ProcessCommand.php (hunk 6 above)
Comment (mssimi) [reply src/Console/Command/ProcessCommand.php:140]: Like I said, we don't need to invalidate files, we override them, same does phpstan.

File: packages/Caching/UnchangedFilesFilter.php (hunk 5 above)
Comment (mssimi) [reply packages/Caching/UnchangedFilesFilter.php:33]: Nowhere, no invalidation is needed.

File: packages/Caching/Detector/ChangedFilesDetector.php (hunk 4 above)
Comment (mssimi) [reply packages/Caching/Detector/ChangedFilesDetector.php:53]: It is not invalidated, inspired by phpstan, when file cache is changed, we analyze that file and then we refresh cache for that file. It works, tested on large code base.

**Comment by mssimi** [comment]
> Removing code without adding test looks very suspicious. It's like removing security layer and all it's tests :)
> 
> Having 2-3 new tests would make this much easier to review.

I have done as much as I could, feel free to keep not working cache in rector, I will use this fork.

As for tests, whole FileCacheStorage was added by you and has no single test, I have added nothing and you want it from me, seems pretty much hypocritical to me.

Anyway, have a nice day.

@staabm Thx for helping me, I really appreciate it, I wish everyone was that helpful like you. :+1:

**Comment by mssimi** [comment]
Closing as I will solve issue with patch using https://github.com/cweagans/composer-patches.

Patch here: https://gist.github.com/mssimi/9cfca094b952f76a3dcb49fd8651bac3
